backend/src/main/scala/cromwell/backend/io/DirectoryFunctions.scala

```diff
@@ -28,8 +29,16 @@ trait DirectoryFunctions extends IoFunctionSet with PathFactory {
     }
   }
 
-  override def isDirectory(path: String) = Future.fromTry(Try(buildPath(path).isDirectory))
+  override def isDirectory(path: String) = asyncIo.isDirectory(buildPath(path))
 
+  /*
+   * Several things are wrong here.
+   * 1) None of this is going through the I/O Actor,
+   * which means no instrumentation, no throttling, no batching, and no custom retries.
+   * 2) The NIO implementation of "list" in GCS will list all objects with the prefix "path", unlike the unix
+   * implementation which lists files and directories children. What we need is the unix behavior, even for cloud filesystems.
+   * 3) It uses the isDirectory function directly on the path, which cannot be trusted for GCS paths. It should use asyncIo.isDirectory instead.
+   */
```

Comment:
🙈

Reply:
All true and perhaps like the other CWL file attributes I'm not sure we should really be doing any of this eagerly.